src/interrupt.cpp

```diff
@@ -0,0 +1,152 @@
+    return requested;
+  }
+ 
+private:
+  InterruptPollingSignal(const InterruptPollingSignal& other); 
+  InterruptPollingSignal& operator=(const InterruptPollingSignal&);
+private:
+  mutex mutex_; 
+  bool pollingRequested_;
+};
+
+InterruptPollingSignal s_pollingSignal;
+
+extern "C" {
+
+// Forward declarations
+int pollForInterrupts(void*);
+void checkUserInterrupt(void*);
+  
+// Background thread which re-schedules pollForInterrupts on the main Python
+// interpreter thread every 100ms so long as the Python interpeter is still
+// running (when it stops running it will stop calling pollForInterrupts and
+// the polling signal will not be set).
+void interruptPollingWorker(void *) {
+  while(true) {
+    
+    // Throttle via sleep
+    this_thread::sleep_for(chrono::milliseconds(100));
+    
+    // Schedule polling on the main thread if the interpeter is still running
+    // Note that Py_AddPendingCall is documented to be callable from a background
+    // thread: "This function doesn’t need a current thread state to run, and it 
+    // doesn’t need the global interpreter lock."
+    // (see: https://docs.python.org/3/c-api/init.html#c.Py_AddPendingCall)
+    if (s_pollingSignal.collectRequest())
+      Py_AddPendingCall(pollForInterrupts, NULL);
+    
+  }
+}
+
+// Callback function scheduled to run on the main Python interpreter loop. This
+// is scheduled using Py_AddPendingCall, which ensures that it is run on the
+// main thread while the interpreter is executing. Note that we can't just have
+// this function keep reschedulding itself or the interpreter would be swamped
+// with just calling and re-calling this function. Rather, we need to throttle
+// the scheduling of the function by using a background thread + a sleep timer.
+int pollForInterrupts(void*) {
+  
+  // Check whether an interrupt has been requested by the user. If one
+  // has then set the Python interrupt flag (which will soon after result
+  // in a KeyboardInterrupt error being thrown).
+  if (R_ToplevelExec(checkUserInterrupt, NULL) == FALSE)
+    PyErr_SetInterrupt();
+  
+  // Request that the background thread schedule us to be called again
+  // (this is delegated to a background thread so that these requests
+  // can be throttled)
+  s_pollingSignal.requestPolling();
+  
+  // Success
+  return 0;
+}
+
+
+// Wrapper for calling R_CheckUserInterrupt within R_TolevelExec. Note that 
+// this call will result in R calling it's internal R_ProcessEvents function 
+// which will allow front-ends to pump events, set the interrupt pending flag, 
+// etc. This function may also jump_to_top, but these jumps are caught by 
+// R_ToplevelExec and results in a return value of FALSE.
+void checkUserInterrupt(void*) {
+  R_CheckUserInterrupt();
+}
+  
+} // extern "C"
+} // anonymous namespace
+
+
+// Initialize interrupt polling background thread
+void initialize_interrupt_polling() {
+  thread t(interruptPollingWorker, NULL);
+  t.detach();
```

Comment:
If Python or R crashes then the whole process is going to crash (either is basically an end of the world scenario). I could put `catch(...)` around the handler code but it's all C code it's calling so there shouldn't be any C++ exceptions (and if there are I'd like to preserve whatever stack trace we can get).